Alignment/OfflineValidation/bin/Zmumumerge.cc

```diff
@@ -0,0 +1,292 @@
+  TDirectoryFile* tdirectory = (TDirectoryFile*)inputfile->Get("myanalysis");
+  for (int i = 0; i < variables_number; i++) {
+    TString th2d_name = Form("th2d_mass_%s", tstring_variables_name[i].Data());
+    th2d_mass_variables[i] = (TH2D*)tdirectory->Get(th2d_name);
+  }
+
+  gSystem->Exec(Form("mkdir -p %s", output_path.Data()));
+  gSystem->Exec(Form("mkdir -p %s/fitResultPlot", output_path.Data()));
+  TFile* outpufile = TFile::Open(Form("%s/fitting_output.root", output_path.Data()), "recreate");
+  TH1D* th1d_variables_meanmass[variables_number];
+  TH1D* th1d_variables_entries[variables_number];
+  const int variables_rebin[variables_number] = {1, 1, 1, 1, 1, 1, 1, 1};
+  const double xaxis_range[variables_number][2] = {
+      {-1, 1}, {-4.8, 4.8}, {-2.4, 2.4}, {-2.4, 2.4}, {-1, 1}, {-M_PI, M_PI}, {-M_PI, M_PI}, {0, 100}};
+  for (int i = 0; i < variables_number; i++) {
+    TString th1d_name = Form("th1d_meanmass_%s", tstring_variables_name[i].Data());
+
+    th2d_mass_variables[i]->RebinY(variables_rebin[i]);
+    th1d_variables_meanmass[i] = th2d_mass_variables[i]->ProjectionY(th1d_name, 1, 1, "d");
+    for (int j = 0; j < th1d_variables_meanmass[i]->GetNbinsX(); j++) {
+      if (i == 7 and j > 25) {
+        continue;
+      }
+      cout << "th1d_variables_meanmass[i]->GetNbinsX()=" << th1d_variables_meanmass[i]->GetNbinsX() << endl;
+      cout << "th2d_mass_variables[i]->GetNbinsY()=" << th2d_mass_variables[i]->GetNbinsY() << endl;
+      th1d_variables_meanmass[i]->SetBinContent(j, 0);
+      th1d_variables_meanmass[i]->SetBinError(j, 0);
+
+      TString th1d_mass_temp_name = Form("th1d_mass_%s_%d", tstring_variables_name[i].Data(), j);
+      TH1D* th1d_i = th2d_mass_variables[i]->ProjectionX(th1d_mass_temp_name, j, j, "d");
+      th1d_i->Write(th1d_mass_temp_name);
+      TString s_cut = Form("nocut");
+      TString s_name = Form("%s_%d", tstring_variables_name[i].Data(), j);
+
+      FitOut fitR = ZMassBinFit_OldTool(th1d_i, s_name, output_path);
+      th1d_variables_meanmass[i]->SetBinContent(j, fitR.mean);
+      th1d_variables_meanmass[i]->SetBinError(j, fitR.mean_err);
+    }
+    th1d_variables_meanmass[i]->GetXaxis()->SetRangeUser(xaxis_range[i][0], xaxis_range[i][1]);
+    Draw_th1d(th1d_variables_meanmass[i], tstring_variables_name[i]);
+    outpufile->cd();
+    th1d_variables_meanmass[i]->Write(th1d_name);
+
+    TString th1d_name_entries = Form("th1d_entries_%s", tstring_variables_name[i].Data());
+    th1d_variables_entries[i] = th2d_mass_variables[i]->ProjectionY(th1d_name_entries, 0, -1, "d");
+    th1d_variables_entries[i]->GetXaxis()->SetTitle(tstring_variables_name[i].Data());
+    th1d_variables_entries[i]->GetYaxis()->SetTitle("Entry");
+    outpufile->cd();
+    th1d_variables_entries[i]->Write(th1d_name_entries);
+  }
+
+  outpufile->Write();
+  outpufile->Close();
+  delete outpufile;
+}
+
+const static int max_file_number = 10;
+void Draw_TH1D_forMultiRootFiles(vector<TString> file_names,
```

Comment:
All these parameters should be more conveniently passed by reference... but ok, this is some "semi-private" validation code: let merge it, and feel free to update further if you think it can be needed,

Also your latest fix does not prevent the possibility of accessing `th1d_input[0]` uninitialized at [L222](https://github.com/cms-sw/cmssw/pull/40346/files#diff-63545e75c9947cc0ce443dc5d66aab3699a2f3462bea640914c9ba1142f6e9e4R222). But still, the whole code wouldn't simply run if your pass it emtpy vectors, therefore I also won't stess it too much.

Reply:
thanks for the extra comments: see https://github.com/cms-sw/cmssw/pull/40371